docs/developer-docs/smart-contracts/advanced-features/async-code.mdx

```diff
@@ -1,90 +1,94 @@
+
+* If the callee is unresponsive (which could happen because of high load on the callee, high load or an error on the callee's subnet, or even because the callee is malicious and delays the response on purpose), the caller canister is stalled and has no control over when it can resume processing or provide an answer. Furthermore, the caller canister cannot be safely stopped and upgraded.
+* They do not scale well. When the system is under high load, canisters may be unable to issue further inter-canister calls.
+
+Best-effort calls address both of these problems. However, they can require more effort on the part of the caller to handle the additional error case. Here are some guidelines how to choose between the two types of calls:
+
+* Always prefer best-effort response calls for calls that don't change the state of the callee, i.e., reads.
+* For endpoints that change the state of the callee, the best practice is to make such endpoints amenable to [safe retries](/docs/current/developer-docs/smart-contracts/best-practices/idempotency), since this is also required for safely handling external user calls. Use best-effort responses calls for such endpoints, and handle the additional edge cases.
+* Use guaranteed response calls for endpoints that mutate state and do not enable safe retries, or calls that perform larger cycle transfers. Best-effort response calls can currently lose cycles in edge cases. Be aware of the limitations of guaranteed response calls listed above. If safe upgrades are needed, consider using a stateless proxy canister (see the [security best practices](https://internetcomputer.org/docs/current/developer-docs/security/security-best-practices/inter-canister-calls) for more information).
+
+## `async/await` syntax, concurrency and state changes
+
+The ``async/await`` syntax allows canister methods to issue asynchronous inter-canister calls and still be structured like an ordinary synchronous function. In this syntax, a method `foo` on a canister `A` could be written something like (in Rust-like syntax):
 
 ```
-/// Canister A
 async fn foo() {
-  baz().await;
+  do_work();
   call(canister_b, 'bar').await;
   do_more_work(res);
 }
+```
 
-async fn baz() {
-  do_local_work();
-}
+While this looks like an ordinary function, there are important differences in behavior between synchronous functions and functions that perform inter-canister calls.
 
-/// Canister B
-fn bar() {
-  do_some_more_work();
-}
-```
+As already mentioned, unlike, say, Ethereum, multiple method call executions on the same canister can be executed concurrently in the presence of inter-canister calls. This increases the canister's throughput, but it also means that the code needs to be correct also in the presence of concurrent behaviors. In particular, the developers have to ensure that no re-entrancy issues occur.
```

Comment:
```suggestion
Multiple method call executions on the same canister can be executed concurrently in the presence of inter-canister calls. This increases the canister's throughput, but it also means that the code needs to be correct also in the presence of concurrent behaviors. In particular, the developers have to ensure that no re-entrancy issues occur.
```

Reply:
Reworded; I want to keep some reference to Ethereum.